### PR TITLE
KBaseSets.SampleSet Search capabilities

### PR DESCRIPTION
**Pull request by slebras**

### CODE REVIEW

**Review by Tianhao-Gu — APPROVED** [review]
👍  I am not too familiar with the search argument for sample set. But the code looks good to me.

File: SetAPI.spec [hunk 1]
```diff
@@ -628,7 +628,26 @@ module SetAPI {
     funcdef create_sample_set(CreateRNASeqSampleSetParams params)
                       returns(CreateRNASeqSampleSetResult) authentication required;
 
+    typedef tuple<string column, boolean ascending> column_sorting;
 
+    typedef structure {
```
Comment (briehl) [inline SetAPI.spec:647]: Some commentary here would be helpful. Are there limits on how many you can request (the "limit" value, I'm guessing has some max)? What column strings are allowed? Is there an expected format for the query string? And so on.

File: lib/SetAPI/sampleset/SampleSearchUtils.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        # we sort by "name" by default
+        sort_by = [("name", 1)]
+        extra_must = []
+        if params.get('sort_by'):
+            sort_by = params['sort_by']
+        if params.get('query'):
+            # use default all_sample_metadata_field as specified in the index_runner spec
+            extra_must.append({"match": {"all_sample_metadata_field": str(params['query'])}})
+        start = params.get('start', 0)
+        limit = params.get('limit', 10)
```
Comment (briehl) [inline lib/SetAPI/sampleset/SampleSearchUtils.py:23]: including these defaults in the documentation would be helpful!

File: test/SampleSetSearch_test.py [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+
+    def _compare_samples(self, s1, s2):
+        self.assertEqual(s1['num_found'], s2['num_found'])
+        self.assertEqual(s1['start'], s2['start'])
+        self.assertEqual(s1['samples'], s2['samples'])
+
+    # @unittest.skip('x')
+    def test_sample_set_to_sample_info(self):
+        ret = self.serviceImpl.sample_set_to_samples_info(self.ctx, {
+            "ref": self.sample_set_ref,
```
Comment (briehl) [inline test/SampleSetSearch_test.py:107]: What happens if you leave out ref? Or start? Or limit? Should probably all be tested for errors.

File: lib/SetAPI/sampleset/SampleSearchUtils.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        elif 'hits' in resp:
+            return {
+                "num_found": int(resp['count']),
+                "start": start,
+                "query": query,
+                "samples": []
+            }
+        else:
+            # only raise if no hits found at highest level.
+            raise RuntimeError(f"no 'hits' with params {json.dumps(params)}\n in http response: {resp}")
```
Comment (briehl) [inline lib/SetAPI/sampleset/SampleSearchUtils.py:96]: Is this a possible error from the service? If it's possible for zero hits, would it return something like `{"hits": {}}` or an error?

File: lib/SetAPI/sampleset/SampleSearchUtils.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            print(f"querying {self.search_url}, with params: {json.dumps(query_data)}")
+        resp = requests.post(self.search_url, headers=headers, data=json.dumps(query_data))
+        if self.debug:
+            print(f"queried {self.search_url} with data {json.dumps(query_data)}")
+        if not resp.ok:
+            raise Exception(f"Not able to complete search request against {self.search_url} "
+                            f"with parameters: \n{json.dumps(query_data)} \nResponse body:\n{resp.text}")
+        respj = resp.json()
+        if self.debug:
+            print(f'respone from query: {respj}')
```
Comment (briehl) [inline lib/SetAPI/sampleset/SampleSearchUtils.py]: typo - response

File: lib/SetAPI/sampleset/SampleSearchUtils.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        if respj.get('error'):
+            raise Exception(f"Error from Search API with parameters\n{json.dumps(query_data)}\n Response body:\n{respj}")
```
Comment (briehl) [inline lib/SetAPI/sampleset/SampleSearchUtils.py]: Might consider a ValueError or RuntimeError instead of just a raw Exception.

**Review by briehl — CHANGES_REQUESTED** [review]
Looks pretty good to me. I have a couple requests though - mainly more documentation in spec files, and some fail checking tests.

File: lib/SetAPI/sampleset/SampleSearchUtils.py (hunk 4 above)
Comment (slebras) [reply lib/SetAPI/sampleset/SampleSearchUtils.py:96]: depends on the type of error. I've had a consistent issue in the past where bad queries don't always result in an error status and instead fail quietly. This condition is to catch those situations

**Pull request merged**